fixes formatting for extremely high B-factor values

OWissett opened the pull request:
<!--- Please read each of the following items and confirm by replacing
 !--the [ ] with a [X] --->

- [x] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [x] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit``
locally, and understand that continuous integration checks will be used to
confirm the Biopython unit tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

<!--- Briefly describe the changes included in this pull request below
 !--- starting with 'Closes #...' if appropriate --->

Closes #5046

JoaoRodrigues commented on Bio/PDB/PDBIO.py:
```diff
@@ -239,7 +242,19 @@ def _get_atom_line(
+                if bfactor < 999
+                else _ATOM_FORMAT_STRING_HIGH_B_FACTOR
+            )
+            if bfactor >= 9999:
```
Quite honestly, instead of throwing a warning and still writing a bad file, I'm more inclined to throwing a warning and truncating the value to a maximum. You can also, instead of having two different atom format strings, which can be annoying to maintain in the future, replace the b-factor `%6.2f` by `%6s` and then make the b-factor conversion when you write it.

Something like:

```python

MAX_B_FACTOR = 9999.9

def format_b_factor(value: float) -> str:
    """Return the b-factor value as a formatted string.
    
    Formats the b-factor value to fit the wwPDB spec of maximum
    6 characters, truncating the value if necessary.
    """
    if value < 1000:  # default
        return f"{value:6.2f}"
    elif value < 10_000:  # modify decimal places
        return f"{value:6.1f}"
    else:  # truncate
        warnings.warn(
            f"Truncated b-factor value {value!r} to {MAX_B_FACTOR} to fit the PDB spec. "
            "Consider writing mmCIF files instead.")
        return str(MAX_B_FACTOR)
```

OWissett commented:
Will do on Monday. I'm away from civilisation this weekend. I'll be up a mountain.

OWissett commented:
I have added unit tests, and restructured how the formatting works a little bit. Since technically the maximum value for a bfactor can be 999_999 (if we have no decimals) we can now handle those without emitting a warning.

I also have accounted for the case where rounding increases the number of digits.

Finally, I added a new warning class `PDBIOWarning`, as I feel it makes sense to have a dedicated warning for saving PDBs compared to other PDB warnings, in the same way we have a `PDBContructionWarning`

JoaoRodrigues commented on Bio/PDB/PDBExceptions.py:
```diff
@@ -28,3 +28,7 @@ class PDBConstructionWarning(BiopythonWarning):
     """Define class PDBIOException."""
+
+
+class PDBIOWarning(BiopythonWarning):
```
Nice, thanks!

JoaoRodrigues commented on Bio/PDB/PDBIO.py:
```diff
@@ -24,6 +22,44 @@
+    """
+    _bfactor_str = ""
+    if value < _MAX_B_FACTOR_2DP:
+        _rounded = round(value, 2)
```
I'm not sure the rounding here is necessary since you're effectively doing it with the string formatting below.

OWissett replied:
Done.

JoaoRodrigues commented on Tests/test_PDB_PDBIO.py:
```diff
@@ -26,6 +26,7 @@
 from Bio.PDB import Select
 from Bio.PDB.PDBExceptions import PDBConstructionWarning
 from Bio.PDB.PDBExceptions import PDBIOException
+from Bio.PDB.PDBIO import _MAX_B_FACTOR, PDBIOWarning
```
I'd import the warning on it's own, like above with `PDBIOException` (and from the Exceptions module)

JoaoRodrigues reviewed (approved):
Minor comments, but looks good to me. If you can, add a small snippet to `NEWS.rst` explaining this change, as this can surprise some people.

JoaoRodrigues commented:
Thanks a lot @OWissett, super nice contribution. I've enabled the full CI workflow to run on these changes and once the tests/checks pass, happy to merge.

JoaoRodrigues commented:
Merged, thanks again @OWissett !
